Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -354,41 +371,58 @@ private class BetterHTTPParser {
     }
 
     func feedInput(_ bytes: UnsafeRawBufferPointer?) throws -> Int {
-        var parserErrno: UInt32 = 0
-        let parserConsumed = self.withExclusiveHTTPParser { parserPtr -> Int in
-            let parserResult: Int
+        var bytesRead = 0
+        let parserErrno: llhttp_errno_t = self.withExclusiveHTTPParser { parserPtr -> llhttp_errno_t in
+            var rc: llhttp_errno_t
+
             if let bytes = bytes {
                 self.rawBytesView = bytes
                 defer {
                     self.rawBytesView = .init(start: UnsafeRawPointer(bitPattern: 0xdafbabe), count: 0)
                 }
-                parserResult = c_nio_http_parser_execute_swift(parserPtr,
-                                                               &self.settings,
-                                                               bytes.baseAddress! + self.httpParserOffset,
-                                                               bytes.count - self.httpParserOffset)
+
+                let startPointer = bytes.baseAddress! + self.httpParserOffset
+                let bytesToRead = bytes.count - self.httpParserOffset
+
+                rc = c_nio_llhttp_execute_swift(parserPtr,
+                                                startPointer,
+                                                bytesToRead)
+
+                if rc == HPE_PAUSED_UPGRADE {
+                    // This is a special pause. We don't need to stop here (our other code will prevent us
+                    // parsing past this point, but we do need a special hook to work out how many bytes were read.
+                    // The force-unwrap is safe: we know we hit an "error".
+                    bytesRead = UnsafeRawPointer(c_nio_llhttp_get_error_pos(parserPtr)!) - startPointer
+                    c_nio_llhttp_resume_after_upgrade(parserPtr)
+                    rc = HPE_OK
```

Comment:
Just for my understanding could you explain me what exactly is going on here and why are we setting `rc` to `HPE_OK`?

Reply:
Sure!

llhttp has a "pause on upgrade" behaviour that will stop the parser at the point that it has completely parse a message that will trigger an upgrade, usually a 101 response. This is sensible: whatever bytes fall behind the parser from the point are not likely to be HTTP and so probably shouldn't be parsed.

In our case, however, we have different logic that handles this "stop parsing" behaviour, because http_parser didn't have the "pause on upgrade" behaviour so we invented it ourself. This essentially mimics the previous behaviour: we work out how many bytes actually got consumed, then we just pretend that everything went great and let the rest of the code kick in.